### PR TITLE
Preliminary addons implementaion with DropBox

### PR DESCRIPTION
**Pull request by chrisseto**

### CODE REVIEW

File: addons/base/models.py [hunk 1]
```diff
@@ -0,0 +1,833 @@
...
+    # TODO: Test me @asmacdo
+    @property
+    def nodes_authorized(self):
+        """Get authorized, non-deleted nodes. Returns an empty list if the
+        attached add-on does not include a node model.
+        """
+        model = self._meta.app_config.node_settings
+        if not model:
+            return []
+        return model.objects.find(user_settings=self, owner__is_deleted=False)
```
Comment (sloria) [inline addons/base/models.py]: Doesn't this need to be `filter`?

File: addons/base/models.py [hunk 2]
```diff
@@ -0,0 +1,833 @@
...
+        pass
+
+    def on_delete(self):
+        """Called when the addon is deleted from the owner (User or Node)."""
+        pass
+
+
+class AddonUserSettingsBase(AddonSettingsBase):
+
+    owner = models.OneToOneField(OSFUser)
```
Comment (sloria) [inline addons/base/models.py]: Do you need to set `null=True`  and `blank=True` on this field?

File: addons/base/models.py [hunk 3]
```diff
@@ -0,0 +1,833 @@
...
+from framework.auth.decorators import must_be_logged_in
+from framework.exceptions import HTTPError
+from framework.exceptions import PermissionsError
+
+from website.addons.base import serializer, logger
+from website.oauth.signals import oauth_complete
+from website.util import waterbutler_url_for
+
+
+class AddonSettingsBase(BaseModel):
```
Comment (sloria) [inline addons/base/models.py]: I think this will need to inherit from `ObjectIDMixin`.

File: addons/base/models.py [hunk 4]
```diff
@@ -0,0 +1,833 @@
...
+        nodes = [Node.load(node_id) for node_id in self.oauth_grants.keys()]
+        for node in nodes:
+            node_addon = node.get_addon(self.oauth_provider.short_name)
+            if node_addon and node_addon.user_settings == self:
+                node_addon.clear_auth()
+
+
+class AddonNodeSettingsBase(AddonSettingsBase):
+
+    owner = models.OneToOneField(Node)
```
Comment (sloria) [inline addons/base/models.py]: Same here; I think this needs to explicitly be set to nullable.

File: addons/base/tests/views.py [hunk 5]
```diff
@@ -0,0 +1,402 @@
+import httplib as http
```
Comment (sloria) [inline addons/base/tests/views.py]: Minor: We might want to move these tests into `osf_models_tests` in the OSF to keep things cohesive.

Will these tests get run by Travis?

File: addons/base/models.py (hunk 4 above)
Comment (chrisseto) [inline addons/base/models.py]: What's the case that owner can be null? Shouldn't that never happen?

Comment (sloria) [inline addons/base/models.py]: I would think so, but [this check](https://github.com/CenterForOpenScience/osf-models/pull/130/files#diff-efa2839fe769cc45f3802492e55481d4R115) gave me reason to doubt.

Comment (sloria) [inline addons/base/models.py]: It's very possible that check is unnecessary.

**Comment by cwisecarver** [comment]
@chrisseto 😍  I made @sloria's changes and pushed up a new PR. I'm about to merge it now.

**Pull request closed without merging**